### PR TITLE
Movie details

### PR DESCRIPTION
**Pull request by TheGuyWhoCodes**
## Majors
 - Add "Movie Details" servlet, users will pass in a movie id and get metadata related to the movie

## Minors
 - Adds in template for config.properties
 - Update gitignore to match for any .properties files

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+     * @throws IOException
+     */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("application/json; charset=utf-8");
+
+        Integer id = null;
+
+        if(null != request.getParameter("id")) {
+            id = Integer.parseInt(request.getParameter("id"));
```
Comment (henryksloan) [inline src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:43]: `Integer.parseInt` can throw NumberFormatException, it might be a good idea to catch it with `response.sendError`.

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        if(null == id) {
```
Comment (henryksloan) [inline src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:46]: Are you sure `Integer.parseInt` can return null? I think it returns a primitive `int` on success, and throws NumberFormatException on failure. I think `id` should be a primitive `int`, an error should be shown if 1) getParameter("id") is null, or 2) Integer.parseInt throws NumberFormatException

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 1 above)
Comment (TheGuyWhoCodes) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:43]: oof. good thought thank you, yeah that would be kinda awkward if someone sent 
```
{
    id: "lol"
}
```

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 2 above)
Comment (henryksloan) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:46]: Edit: I know the program could theoretically get here if the parameter is null, but the solution I suggested would capture both empty and erroneous Id's.

Comment (TheGuyWhoCodes) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:46]: you're 100% right, thanks!

File: pom.xml [hunk 3]
```diff
@@ -68,6 +68,13 @@
       <artifactId>themoviedbapi</artifactId>
       <version>1.10</version>
     </dependency>
+
```
Comment (kirantumkur) [inline pom.xml:71]: Is it possible to add a comment explaining what this library is doing?

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import info.movito.themoviedbapi.TmdbMovies;
+import info.movito.themoviedbapi.model.MovieDb;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+/**
```
Comment (kirantumkur) [inline src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:15]: Lets remove this, as we get this from the pr, anyways.

File: pom.xml [hunk 5]
```diff
@@ -68,6 +68,13 @@
       <artifactId>themoviedbapi</artifactId>
```
Comment (kirantumkur) [inline pom.xml:68]: Lets add a comment, saying that this library is calling OMDB APIs

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        if(null != request.getParameter("id")) {
+            id = parseInt(request.getParameter("id"));
+            if(null == id) {
+                response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+                return;
+            }
+        } else {
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+            return;
+        }
```
Comment (kirantumkur) [inline src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:48]: Can this be rewritten as?  

 value = request.getParameter("id")
   if value == null {
        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
            return;
   }
 
   id = parseInt(value)
   if id == null {
      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
            return;
   }

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+        response.getWriter().println(gson.toJsonTree(queryResponse));
+    }
+
+    /**
+     * parseInt method is used to parse an integer from a string (from a url param)
+     * @param urlParam: the url param to convert from string -> integer
+     * @return either an integer if the string is a valid integer, or null if it can't be parsed
+     */
+    private Integer parseInt(String urlParam) {
```
Comment (kirantumkur) [inline src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:61]: rename urlParam to value or something meaningful to this function. A good idea is to reuse what ever Integer.parseInt() uses..

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        try {
+            return Integer.parseInt(urlParam);
+        } catch(NumberFormatException e) {
```
Comment (kirantumkur) [inline src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:64]: What if

File: src/test/java/com/google/sps/servlets/movie/MovieSearchServletTest.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+/**
```
Comment (kirantumkur) [inline src/test/java/com/google/sps/servlets/movie/MovieSearchServletTest.java:23]: Please remove "author" from the code everywhere. Code is owned and edited by the whole team, and adding "author" like this makes people avoid wanting to code in areas "owned" by others.

File: src/main/resources/config.properties [hunk 10]
```diff
@@ -0,0 +1 @@
+movie_apikey=insert_here
```
Comment (kirantumkur) [inline src/main/resources/config.properties:1]: For now, please add a comment in this file, which explains how the key is generated and shared between team members. We can follow up with a solution later.

File: .gitignore [hunk 11]
```diff
@@ -88,4 +88,6 @@ npm-debug.log
 *.log
 *.stackdump
 scripts/docs.angularjs.org-firebase/functions/content
-/firebase.json
\ No newline at end of file
+/firebase.json
+
+*.properties
```
Comment (kirantumkur) [inline .gitignore:93]: Ideally, this should go into a separate PR.

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 6 above)
Comment (TheGuyWhoCodes) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:48]: Is rewriting as this fine?

```
        Integer movieId = parseInt(request.getParameter("id"));

        if (movieId == null) {
            response.sendError(HttpServletResponse.SC_BAD_REQUEST);
            return;
        }
```

The parseInt() function handles any null returned by request.getParameter("id") so we can consolidate it into one if block.

File: src/main/resources/config.properties (hunk 10 above)
Comment (TheGuyWhoCodes) [reply src/main/resources/config.properties]: Fixed with 974e7117

File: src/test/java/com/google/sps/servlets/movie/MovieSearchServletTest.java (hunk 9 above)
Comment (TheGuyWhoCodes) [reply src/test/java/com/google/sps/servlets/movie/MovieSearchServletTest.java]: Fixed with a70f4e0d

File: pom.xml (hunk 5 above)
Comment (TheGuyWhoCodes) [reply pom.xml:69]: Fixed with 631568ce

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 4 above)
Comment (TheGuyWhoCodes) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:15]: Fixed with a70f4e0d

File: pom.xml (hunk 3 above)
Comment (TheGuyWhoCodes) [reply pom.xml:72]: Fixed with 631568ce

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 8 above)
Comment (TheGuyWhoCodes) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:64]:
```
        try {
            return Integer.parseInt(urlParam);
        } catch(Exception e) {
            return null;
        }
```

So were you looking for something like this? I looked through `Integer.parseInt()` and I couldn't find any other instances from an error thrown. The method signature is: 
`    public static int parseUnsignedInt(String s, int radix)
                throws NumberFormatException`

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 6 above)
Comment (kirantumkur) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:48]: Yes. New code makes more sense

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 8 above)
Comment (kirantumkur) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:64]: Yes, this makes more sense as "all" exceptions are being handled the same way.

Comment (henryksloan) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:64]: https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#parseInt(java.lang.String)
I believe this particular function can only ever throw NumberFormatException. My understanding is that functions have to declare all exceptions they could throw, so I'm pretty sure catching NumberFormatException is more specific, and will never crash. Correct me if I'm wrong.

Comment (kirantumkur) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:64]: Normally, you don't catch "specific" exceptions. You only catch "specific" exceptions, if you have a "specific" action for that type of exception. Or you just catch all exceptions apply the default handler.

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 7 above)
Comment (kirantumkur) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:61]: Eg: Integer.partInt() is using (string s) as the input. urlParam doesn't make sense, inside of parseInt.

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 8 above)
Comment (henryksloan) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java:64]: Good point, you're right.

File: .gitignore (hunk 11 above)
Comment (TheGuyWhoCodes) [reply .gitignore]: Fixed with 0d8a5d1

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 8 above)
Comment (TheGuyWhoCodes) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java]: Fixed with cc10931

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 6 above)
Comment (TheGuyWhoCodes) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java]: Fixed with 131a2a94

File: src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java (hunk 7 above)
Comment (TheGuyWhoCodes) [reply src/main/java/com/google/sps/servlets/movie/MovieDetailsServlet.java]: Fixed with 131a2a94
